### PR TITLE
feat: Improved Workflow logs

### PR DESCRIPTION
**Pull request by ghawk1ns**
By default, workflow logs for a run will log out run status and individual task status.
Tasks logs can be emitted with `--task <taskId>` for a single task log, `--all-tasks` for all task logs, and `--failed-tasks` for failed task logs.

Issue #, if available:

**Description of Changes**

[//]: #  (A description of the change that you made and the new user experience that it creates)

**Description of how you validated changes**
Verified commands manually 

```
$ agc logs workflow foo
2021-11-09T17:17:01-08:00 𝒊  Showing the logs for 'foo'
2021-11-09T17:17:02-08:00 𝒊  Showing logs for the latest run of the workflow. Run id: '4927d57c-5181-4505-9e9f-96a7590ea9eb'
RunId: 4927d57c-5181-4505-9e9f-96a7590ea9eb
State: COMPLETE
Tasks: 
        Name                    JobId                                   StartTime                               StopTime                                ExitCode
        hello_agc.goodbye       29fc9be3-3cc9-4e44-bbb1-50072b7be053    2021-11-09 20:21:10.118 +0000 UTC       2021-11-09 20:27:01.382 +0000 UTC       0
        hello_agc.hello         d6dad1b4-52ec-4765-8d8a-0f42093d49d2    2021-11-09 20:21:10.117 +0000 UTC       2021-11-09 20:26:31.345 +0000 UTC       0
        hello_agc.no            1764a32d-b64b-4c9f-b5a9-e86df0c1fca7    2021-11-09 20:21:10.117 +0000 UTC       2021-11-09 20:27:10.345 +0000 UTC       0
```

```
$ agc logs workflow foo --all-tasks
2021-11-09T17:17:07-08:00 𝒊  Showing the logs for 'foo'
2021-11-09T17:17:08-08:00 𝒊  Showing logs for the latest run of the workflow. Run id: '4927d57c-5181-4505-9e9f-96a7590ea9eb'
Tue, 09 Nov 2021 12:26:14 -0800 download: s3://agc-388165786712-us-west-
[...]
Tue, 09 Nov 2021 12:26:17 -0800 *** COMPLETED DELOCALIZATION ***
Tue, 09 Nov 2021 12:26:17 -0800 *** EXITING WITH RETURN CODE ***
Tue, 09 Nov 2021 12:26:17 -0800 0
```

```
$ agc logs workflow foo --failed-tasks
2021-11-09T17:17:15-08:00 𝒊  Showing the logs for 'foo'
2021-11-09T17:17:15-08:00 𝒊  Showing logs for the latest run of the workflow. Run id: '4927d57c-5181-4505-9e9f-96a7590ea9eb'
2021-11-09T17:17:16-08:00 𝒊  No logs available for run '4927d57c-5181-4505-9e9f-96a7590ea9eb'. Please try again later.
```

```
$ agc logs workflow foo --task 29fc9be3-3cc9-4e44-bbb1-50072b7be053
2021-11-09T17:17:29-08:00 𝒊  Showing the logs for 'foo'
2021-11-09T17:17:30-08:00 𝒊  Showing logs for the latest run of the workflow. Run id: '4927d57c-5181-4505-9e9f-96a7590ea9eb'
Tue, 09 Nov 2021 12:26:14 -0800 download: s3://agc-388165786712-us-west-1/scripts/69070f57f6499609f65566ed33421740 to tmp/tmp.jCuIQYqtS/batch-file-temp
Tue, 09 Nov 2021 12:26:14 -0800 *** LOCALIZING INPUTS ***
Tue, 09 Nov 2021 12:26:16 -0800 download: s3://agc-388165786712-us-west-1/project/myProject/userid/guyhawki2Xarph/context/groot/cromwell-execution/hello_agc/4927d57c-5181-4505-9e9f-96a7590ea9eb/call-goodbye/script to agc-388165786712-us-west-1/project/myProject/userid/guyhawki2Xarph/context/groot/cromwell-execution/hello_agc/4927d57c-5181-4505-9e9f-96a7590ea9eb/call-goodbye/script
Tue, 09 Nov 2021 12:26:17 -0800 *** COMPLETED LOCALIZATION ***
Tue, 09 Nov 2021 12:26:17 -0800 Hello Amazon Genomics CLI!3
Tue, 09 Nov 2021 12:26:17 -0800 *** DELOCALIZING OUTPUTS ***
Tue, 09 Nov 2021 12:26:18 -0800 upload: ./goodbye-rc.txt to s3://agc-388165786712-us-west-1/project/myProject/userid/guyhawki2Xarph/context/groot/cromwell-execution/hello_agc/4927d57c-5181-4505-9e9f-96a7590ea9eb/call-goodbye/goodbye-rc.txt
Tue, 09 Nov 2021 12:26:19 -0800 upload: ./goodbye-stderr.log to s3://agc-388165786712-us-west-1/project/myProject/userid/guyhawki2Xarph/context/groot/cromwell-execution/hello_agc/4927d57c-5181-4505-9e9f-96a7590ea9eb/call-goodbye/goodbye-stderr.log
Tue, 09 Nov 2021 12:26:20 -0800 upload: ./goodbye-stdout.log to s3://agc-388165786712-us-west-1/project/myProject/userid/guyhawki2Xarph/context/groot/cromwell-execution/hello_agc/4927d57c-5181-4505-9e9f-96a7590ea9eb/call-goodbye/goodbye-stdout.log
Tue, 09 Nov 2021 12:26:20 -0800 *** COMPLETED DELOCALIZATION ***
Tue, 09 Nov 2021 12:26:20 -0800 *** EXITING WITH RETURN CODE ***
Tue, 09 Nov 2021 12:26:20 -0800 0
```

**Checklist**

- [x] If this change would make any existing documentation invalid, I have included those updates within this PR
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [x] I have linted my code before raising the PR
- [x] Title of this Pull Request follows Conventional Commits standard: https://www.conventionalcommits.org/en/v1.0.0/

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

```
go test -race -cover -count=1 -coverprofile coverage.out ./internal...
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws     2.966s  coverage: 93.3% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/batch       1.699s  coverage: 90.9% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/cdk 2.415s  coverage: 68.8% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/cfn 1.211s  coverage: 87.7% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/cwl 0.488s  coverage: 92.2% of statements
?       github.com/aws/amazon-genomics-cli/internal/pkg/aws/ddb [no test files]
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/ecr 0.742s  coverage: 81.8% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/s3  2.531s  coverage: 95.9% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/ssm 2.254s  coverage: 90.0% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/sts 1.463s  coverage: 85.7% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/aws/util        0.953s  coverage: 100.0% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli     3.338s  coverage: 48.2% of statements
?       github.com/aws/amazon-genomics-cli/internal/pkg/cli/awsresources        [no test files]
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/clierror    2.310s  coverage: 58.3% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/clierror/actionableerror    3.023s  coverage: 100.0% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/config      3.108s  coverage: 42.4% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/context     2.971s  coverage: 85.5% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/format      2.996s  coverage: 84.0% of statements
?       github.com/aws/amazon-genomics-cli/internal/pkg/cli/group       [no test files]
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/spec        3.667s  coverage: 52.0% of statements
?       github.com/aws/amazon-genomics-cli/internal/pkg/cli/types       [no test files]
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/workflow    3.414s  coverage: 83.7% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/cli/zipfile     3.528s  coverage: 86.7% of statements
?       github.com/aws/amazon-genomics-cli/internal/pkg/constants       [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/environment     [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/logging [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/mocks/aws       [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/mocks/context   [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/mocks/io        [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/mocks/manager   [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/mocks/storage   [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/mocks/wes       [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/mocks/workflow  [no test files]
ok      github.com/aws/amazon-genomics-cli/internal/pkg/slices  3.623s  coverage: 100.0% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/storage 3.700s  coverage: 75.3% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/term/color      3.739s  coverage: 100.0% of statements
ok      github.com/aws/amazon-genomics-cli/internal/pkg/version 3.284s  coverage: 89.8% of statements
?       github.com/aws/amazon-genomics-cli/internal/pkg/wes     [no test files]
?       github.com/aws/amazon-genomics-cli/internal/pkg/wes/option      [no test files]
(cd packages/cli; /Library/Developer/CommandLineTools/usr/bin/make build)
go build -ldflags "-X github.com/aws/amazon-genomics-cli/internal/pkg/version.Version=1.0.1-65-gdca0fdd" -o ./bin/local/agc ./cmd/application
```

### CODE REVIEW

**Review by IllyaYalovyy — COMMENTED** [review]
Could you please fill out "Description of how you validated changes" section of the PR?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -48,7 +48,7 @@ jobs:
           working-directory: packages/cli/
 
           # Optional: golangci-lint command line arguments.
-          # args: --issues-exit-code=0
+          args: --timeout=5m0s
```
Comment (IllyaYalovyy) [inline .github/workflows/ci.yml:51]: How is this change related to workflow logs?

File: packages/cli/internal/pkg/cli/logs_workflow.go [hunk 2]
```diff
@@ -72,11 +82,40 @@ func (o *logsWorkflowOpts) Execute() error {
...
-	jobIds, err := o.getJobIds()
+
+	runLog, err := o.workflowManager.GetRunLog(o.runId)
 	if err != nil {
 		return err
 	}
 
+	var jobIds []string
+	if o.taskId != "" {
+		var taskExists = false
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/cli/logs_workflow.go:93]: nit. `taskExists := false` or `var taskExists bool`

File: packages/cli/internal/pkg/cli/logs_workflow.go [hunk 3]
```diff
@@ -72,11 +82,40 @@ func (o *logsWorkflowOpts) Execute() error {
...
+		for _, task := range runLog.Tasks {
+			if task.JobId == o.taskId {
+				taskExists = true
+				break
+			}
+		}
+
+		if !taskExists {
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/cli/logs_workflow.go:101]: if !containsTaskId(runLog.Tasks, o.taskId) {
...
}

File: packages/cli/internal/pkg/cli/logs_workflow.go [hunk 4]
```diff
@@ -72,11 +82,40 @@ func (o *logsWorkflowOpts) Execute() error {
...
+		jobIds = []string{o.taskId}
+	} else if o.allTasks || o.failedTasks {
+		jobIds, err = o.getJobIds(runLog.Tasks)
+		if err != nil {
+			return err
+		}
+	} else {
+		b := bytes.NewBufferString("")
+		format.NewStringFormatter(b).Write(runLog)
+		printLn(b)
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/cli/logs_workflow.go:115]: Could you please clarify what this code does? Can it be simplified?

File: packages/cli/internal/pkg/cli/workflow/workflow_tasks.go [hunk 5]
```diff
@@ -32,6 +38,23 @@ func (m *Manager) GetWorkflowTasks(runId string) ([]Task, error) {
 
 }
 
+func (m *Manager) GetRunLog(runId string) (RunLog, error) {
+
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/cli/workflow/workflow_tasks.go:42]: for a manger's method it is a must to check error first thing, and exit if it's not `nil`.

File: packages/cli/internal/pkg/cli/workflow/workflow_tasks.go [hunk 6]
```diff
@@ -32,6 +38,23 @@ func (m *Manager) GetWorkflowTasks(runId string) ([]Task, error) {
...
+	tasks, err := m.GetWorkflowTasks(runId)
+	if err != nil {
+		return RunLog{}, err
```
Comment (IllyaYalovyy) [inline packages/cli/internal/pkg/cli/workflow/workflow_tasks.go:45]: manager's `err` has to be set to this error value. Please see other method of this manager for an example of this design pattern. 

The whole implementation doesn't feel like it belongs here. Please consider different option.

**Review by IllyaYalovyy — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ghawk1ns) [reply .github/workflows/ci.yml:51]: For whatever reason, the linter is sometimes taking more than a minute to pass. I don't think it's related to WF logs, but if it's not increased it will lead to infrequent build failures.

File: packages/cli/internal/pkg/cli/workflow/workflow_tasks.go (hunk 6 above)
Comment (ghawk1ns) [reply packages/cli/internal/pkg/cli/workflow/workflow_tasks.go:45]: > The whole implementation doesn't feel like it belongs here

Can you elaborate? The other calls here are fetching and providing data from the same source (`WES.getRunLog`), for instance `manager.GetWorkflowTasks` provides a subset of RunLog data.

File: packages/cli/internal/pkg/cli/logs_workflow.go (hunk 4 above)
Comment (ghawk1ns) [reply packages/cli/internal/pkg/cli/logs_workflow.go:115]: It's getting a formatted version of the RunLog struct and printing it out with the existing log print logic. If we don't use `printLn` to feed stdout for logs, it would require additional changes to our unit testing. Since that's how we validate log output

Comment (IllyaYalovyy) [reply packages/cli/internal/pkg/cli/logs_workflow.go]: My concern here is that "formatted RunLog struct" is not a log, so "log printing logic" is not applicable to it. I think it should be printed in a similar to all other command results way.

File: packages/cli/internal/pkg/cli/workflow/workflow_tasks.go (hunk 6 above)
Comment (IllyaYalovyy) [reply packages/cli/internal/pkg/cli/workflow/workflow_tasks.go]: So a manager is used, when one have to execute multiple related commands and handle errors behind the scene. This method/use case doesn't fit this pattern. Our managers are already overly complicated and require some refactoring. So I would advise against putting random methods to it. By random, I mean a method doesn't follow the pattern, doesn't execute multiple commands that can return errors.

**Review by abassyiouni — APPROVED** [review]

**Review by IllyaYalovyy — APPROVED** [review]

**Comment by ghawk1ns** [comment]
resolves https://github.com/aws/amazon-genomics-cli/issues/121
